Objects/typeobject.c

```diff
@@ -5728,6 +5728,50 @@ _PyTypes_AfterFork(void)
 #endif
 }
 
+#ifdef Py_GIL_DISABLED
+// Non-caching version of type lookup, containing a non-locking
+// version of find_name_in_mro().
+static unsigned int
+type_lookup_non_interned(PyTypeObject *type, PyObject *name, _PyStackRef *out)
+{
+    PyObject *mro = NULL;
+    Py_hash_t hash = _PyObject_HashFast(name);
+    if (hash == -1) {
+        goto error;
+    }
+
+    /* Keep a strong reference to mro because type->tp_mro can be replaced
+       during dict lookup, e.g. when comparing to non-string keys. */
+    mro = _PyType_GetMRO(type);
```

Comment:
Won't this lead to reference count contention?

I don't think we need to skip the cache in this case. The names are exact unicode objects, they're just not interned. If the string isn't interned can we use a different loop in the lookup code that uses something like `PyUnicode_Compare`?

Reply:
I implemented something that uses unicode compare in gh-133669.  More notes there.